File: ScrollToGradientNavColor/ScrollToGradientNavColor/ViewController.m
//
//  ViewController.m
//  ScrollToGradientNavColor
//
//  Created by allthings_LuYD on 16/8/8.//
//  Copyright © 2016年 scrum_snail. All rights reserved.
//

#import "ViewController.h"
#define SCREEN_WIDTH [UIScreen mainScreen].bounds.size.width
#define SCREEN_HEIGHT [UIScreen mainScreen].bounds.size.height
@interface ViewController ()<UITableViewDelegate,UITableViewDataSource>{
    UIImageView * _topImageView;
}
@property (nonatomic,strong) UITableView *tableView;
@property (nonatomic, assign) CGFloat topContentInset;

@end

@implementation ViewController

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    self.automaticallyAdjustsScrollViewInsets = NO;
    self.navigationController.navigationBar.barTintColor = [UIColor orangeColor];
    self.navigationController.navigationBar.barStyle = UIBarStyleDefault;
    self.navigationController.navigationBar.subviews[0].alpha = 0;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationItem.rightBarButtonItem = self.editButtonItem;
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemReply target:nil action:nil];
    _topContentInset = 136;
    [self.view addSubview:self.tableView];
    [self belowTheTableView];
}

- (void)belowTheTableView{
    _topImageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, SCREEN_WIDTH, SCREEN_WIDTH*871/828.0)];
    _topImageView.image = [UIImage imageNamed:@"backImage"];
    [self.view insertSubview:_topImageView belowSubview:self.tableView];
}

- (UITableView *)tableView{
    if (!_tableView) {
        _tableView = ({
            _tableView = [[UITableView alloc] initWithFrame:self.view.bounds];
            _tableView.delegate = self;
            _tableView.dataSource = self;
            _tableView.contentInset = UIEdgeInsetsMake(64+_topContentInset, 0, 0, 0);
            _tableView.backgroundColor = [UIColor clearColor];
            _tableView;
        });
    }
    return _tableView;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return 50;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *identifier = @"cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
    }
    cell.textLabel.text = @"scrum_snail@yeah.net";

    return cell;
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView{
    CGFloat offsetY = scrollView.contentOffset.y + self.tableView.contentInset.top;
    NSLog(@"%f",offsetY);
    if (offsetY < 0) {
        _topImageView.transform = CGAffineTransformMakeScale(1 + offsetY/(-500), 1 + offsetY/(-500));
    }
    if (offsetY<=_topContentInset*2 && offsetY>(_topContentInset - 64)) {
        CGFloat alpa = offsetY >= _topContentInset *2 ? 1 : offsetY/(_topContentInset * 2);
        self.navigationController.navigationBar.subviews[0].alpha = alpa;
    }
    if (offsetY <= (_topContentInset - 64)) {
        self.navigationController.navigationBar.subviews[0].alpha = 0;
    }
    CGRect rect = _topImageView.frame;
    rect.origin.y = 0;
    _topImageView.frame = rect;
}
@end

File: ScrollToGradientNavColor.swift/ScrollToGradientNavColor.swift/ViewController.swift
//
//  ViewController.swift
//  ScrollToGradientNavColor.swift
//
//  Created by allthings_LuYD on 16/8/8.
//  Copyright © 2016年 scrum_snail. All rights reserved.
//

import UIKit

class ViewController: UIViewController,UITableViewDelegate,UITableViewDataSource{
    let SCREEN_WIDTH = UIScreen.main().bounds.size.width
    let tableView = UITableView()
    let identifier = "cell"
    let topConentInset = 136
    let topImageView = UIImageView()


    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.automaticallyAdjustsScrollViewInsets = false
        self.navigationController?.navigationBar.barTintColor = UIColor.orange()
        self.navigationController?.navigationBar.barStyle = UIBarStyle.default
        self.navigationController?.navigationBar.subviews[0].alpha = 0
    }
    override func viewDidLoad() {
        super.viewDidLoad()
        self.navigationItem.rightBarButtonItem = self.editButtonItem()
        self.navigationItem.leftBarButtonItem = UIBarButtonItem.init(barButtonSystemItem: UIBarButtonSystemItem.reply, target: nil, action: nil)
        setUpUI()
        topImageViewSet()
    }

    func setUpUI() -> Void {
        tableView.frame = self.view.bounds
        tableView.contentInset.top = CGFloat (topConentInset+64)
        tableView.delegate = self
        tableView.dataSource = self
        tableView.backgroundColor = UIColor.clear()
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: identifier)
        self.view.addSubview(tableView)
    }

    func topImageViewSet()  {
        topImageView.frame = CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: SCREEN_WIDTH*871/828.0)
        topImageView.image = UIImage(named: "backImage")
        self.view.insertSubview(topImageView, belowSubview: tableView)
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 50
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: identifier,for: indexPath)
        cell.textLabel?.text = "scrum_snail@yeah.net"
        return cell
    }

    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        let offsetY = scrollView.contentOffset.y + tableView.contentInset.top
        print(offsetY)
        if offsetY < 0 {
            topImageView.transform = CGAffineTransform.init(scaleX: 1+offsetY/(-500), y: 1+offsetY/(-500))
        }
        if offsetY <= CGFloat(topConentInset*2) && offsetY > CGFloat(topConentInset - 64) {
            let alpa = offsetY >= CGFloat(topConentInset*2) ? 1 : offsetY / CGFloat(topConentInset * 2);
            self.navigationController?.navigationBar.subviews[0].alpha = alpa;
        }
        if (offsetY <= CGFloat(topConentInset - 64)) {
            self.navigationController?.navigationBar.subviews[0].alpha = 0;
        }

        topImageView.frame.origin.y = 0
    }

}

